### PR TITLE
Compound integration with Omen Markets. Archived in favor of https://github.com/protofire/omen-exchange/pull/1570 

### PR DESCRIPTION
**Pull request by pRoy24**
Fixed the previous round of comments.
Added badge as defined in the mockup.
Refactored to make it work with ETH as the base and cETH as the underlying collateral.
Tested on Rinkeby with ETH -> cETH and dai -> cDAI pairs.

Please note that you need to run yarn install since this feature uses Big.js version 6.0.0

### CODE REVIEW

**Comment by pimato** [comment]
Hey @pRoy24 and thanks for submitting your Pull Request. There are several improvements we can make to improve the user experience but overall really great work!

**Comment by pimato** [comment]
I think it is important to add an indicator to the market details view that this market is actually earning interest via compound. This is my first attempt for it:

Details

**Figma**: 
[Design Specification](https://www.figma.com/file/R056tSYMjvfwaFidEBItYy/Widgets-V2?node-id=1209%3A141)

File: app/src/components/market/common/add_compound_service/index.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  -webkit-box-align: center;
+`
+
+const ServiceTextWrapper = styled.div`
+  width: 90%;
+`
+
+const ServiceCheckWrapper = styled.div`
+  width: 10%;
+`
```
Comment (Mi-Lan) [inline app/src/components/market/common/add_compound_service/index.tsx:64]: wondering if we could do without these two classes?

File: app/src/components/market/common/add_compound_service/index.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                </div>
+              </ServiceTextWrapper>
+              <ServiceCheckWrapper onClick={toggleServiceCheck}>
+                <CheckService>{serviceChecked}</CheckService>
+              </ServiceCheckWrapper>
+            </ServiceTokenDetails>
+          </ServiceWrapper>
+        </Description>
+      </DescriptionWrapper>
+    </Wrapper>
```
Comment (Mi-Lan) [inline app/src/components/market/common/add_compound_service/index.tsx:116]: Maybe we can check what @KadenZipfel  thinks since he is more senior developer but i believe this whole section has too many unnecessary classes and could be reduced...

File: app/src/components/market/common/add_compound_service/index.tsx [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            <ServiceTokenDetails>
+              <ServiceTextWrapper>
+                <div>Compound</div>
+                <div>
+                  Convert {currentToken} to c{currentToken} to earn {compoundInterestRate}% interest
+                </div>
+              </ServiceTextWrapper>
+              <ServiceCheckWrapper onClick={toggleServiceCheck}>
+                <CheckService>{serviceChecked}</CheckService>
+              </ServiceCheckWrapper>
```
Comment (Mi-Lan) [inline app/src/components/market/common/add_compound_service/index.tsx:111]: Screenshot 2021-01-05 at 13 08 55
Seems that this button is not center aligned. Also when it's clicked it should invert colors but now the way it is now checkmark just disappears...

File: app/src/components/market/sections/market_buy/market_buy.tsx [hunk 4]
```diff
@@ -256,14 +374,16 @@ const MarketBuyWrapper: React.FC<Props> = (props: Props) => {
...
+              disabled={false}
+              filters={filters}
               onSelect={(token: Token | null) => {
                 if (token) {
-                  setCollateral(token)
+                  setBuyCollateral(token)
                   setAmount(new BigNumber(0))
+                  setDisplayAmountToFund(new BigNumber(0))
                 }
               }}
```
Comment (Mi-Lan) [inline app/src/components/market/sections/market_buy/market_buy.tsx:388]: Screenshot 2021-01-05 at 12 45 49
I seem to be getting this error at random time? What is that related to?

**Review by Mi-Lan — CHANGES_REQUESTED** [review]
Hey men good job navigating our codebase seems that you have gotten the hang of it pretty fast! Also i'm having issue creating market? Do you know whats the problem here is the video...But regular markets work good i can create them...

https://user-images.githubusercontent.com/33226956/103646307-c1f9f680-4f59-11eb-8dcd-fbec8545fd8b.mov

File: app/src/components/market/sections/market_sell/market_sell.tsx [hunk 5]
```diff
@@ -38,6 +49,34 @@ const StyledButtonContainer = styled(ButtonContainer)`
...
+  }
+`
+const CurrencyDropdownLabelContainer = styled.div`
+  margin-top: 20px;
+`
+const CurrencyDropdownLabel = styled.div`
+  display: inline-flex;
+  padding-left: 14px;
+  padding-top: 14px;
+  color: #37474f;
```
Comment (KadenZipfel) [inline app/src/components/market/sections/market_sell/market_sell.tsx:74]: Color from `themes.js` here too

File: app/src/services/compound_service.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    this.provider = provider
+  }
+
+  init = async () => {
+    this.exchangeRate = await this.calculateExchangeRate()
+  }
+
+  calculateSupplyRateAPY = async (): Promise<number> => {
+    const supplyRate: number = await this.contract.supplyRatePerBlock()
+    const ethMantissa = 1e18
```
Comment (KadenZipfel) [inline app/src/services/compound_service.ts:32]: Should the mantissa be different for different token decimal amounts?

**Review by KadenZipfel — CHANGES_REQUESTED** [review]
This is quite a big change, and overall it's pretty well done. Some additional notes:

- `batch transaction execution` fails at market creation if compound service is enabled.
- `list_view` displays the cToken as the currency but elsewhere it's abstracted away.
    - 
    - I'm assuming it should also be abstracted away here as well.
- Error depositing DAI in DAI market with compound service enabled:
    - 
    - cDai deposit seems to work fine though
- Base token balance is sometimes incorrectly set to 0
    - 
- Outcome probabilities sometimes get messed up in the buy view
    - 
- It seems like there may be some reused logic throughout components, wherever possible, we should make reusable functions in `tools.ts`

**Review by pimato — CHANGES_REQUESTED** [review]
on funding page loaded, app freezes

File: app/src/components/market/common/add_compound_service/index.tsx (hunk 1 above)
Comment (pRoy24) [reply app/src/components/market/common/add_compound_service/index.tsx:66]: I think this is needed to enforce the width of the respective containers

File: app/src/components/market/sections/market_buy/market_buy.tsx (hunk 4 above)
Comment (pRoy24) [reply app/src/components/market/sections/market_buy/market_buy.tsx:421]: This shoudl be resolved now

File: app/src/services/compound_abi.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/* eslint no-use-before-define: 0 */  // --> OFF
```
[the remark this answers is not in the document]
Comment (pRoy24) [reply app/src/services/compound_abi.ts:1]: Moved the file to abi directory, I think prettifying makes the file larger and since we are not planning to edit the ABIs anyways, maybe just leave it as is and replace it with newer ABIs as Compound release new upgrades.

File: app/src/services/compound_service.ts (hunk 6 above)
Comment (pRoy24) [reply app/src/services/compound_service.ts:37]: yeah I was a bit confused about this too since the docs have an example for ETH but the 1e18 mantissa seems to give the correct APY for all tokens regardless of the decimals so just changed the name of the constant to make it more generic but left the value as is.

**Comment by pimato** [comment]
Just got this error:
Bildschirmfoto 2021-01-11 um 15 02 47

For this market:
Bildschirmfoto 2021-01-11 um 15 04 14

**Review by pimato — CHANGES_REQUESTED** [review]
see comment.

File: app/src/services/cpk.ts [hunk 8]
```diff
@@ -280,6 +347,40 @@ class CPKService {
...
+            value: marketData.funding.toString(),
+          })
+        } else {
+          collateral = marketData.collateral
+          // For any other compound pair that is not ETH
+          const encodedMintFunction = CompoundService.encodeMintTokens(
+            compoundTokenDetails.symbol,
+            marketData.funding.toString(),
+          )
+          // Approve cToken for the cpk contract
```
Comment (hexyls) [inline app/src/services/cpk.ts:373]: I had trouble creating markets with DAI, I think we might need to add an extra step here: Transfer the underlying token from the user to the CPK before we mint the cToken.

File: app/src/components/market/common/add_compound_service/index.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+const TextBody = 'div'
+
+export interface AddCompoundServiceProps {
+  isServiceChecked: boolean
+  toggleServiceCheck?: any
+  compoundInterestRate: string
+  currentToken: string
+}
+
+export const AddCompoundService: React.FC<AddCompoundServiceProps> = (props: AddCompoundServiceProps) => {
```
Comment (hexyls) [inline app/src/components/market/common/add_compound_service/index.tsx:94]: At the moment the compound service will crash the app on Sokol and xDAI networks, it would be awesome if you could disabled the service when connected to these networks 👍

Here are some details on how to setup [Sokol](https://github.com/protofire/omen-exchange/pull/1448) and [xDAI](https://github.com/protofire/omen-exchange/pull/1464) on metamask.

**Review by hexyls — CHANGES_REQUESTED** [review]
Awesome work so far! This is a huge, complicated feature that is coming together nicely 🎉

**Comment by pimato** [comment]
Hey @pRoy24 nice work!

Just a few UI related comments:

Remove margin-left here:
Bildschirmfoto 2021-01-13 um 22 22 41

**Comment by pimato** [comment]
Remove margin-left here:
Bildschirmfoto 2021-01-13 um 22 22 41

**Comment by pimato** [comment]
only `earn 2.64% APY` should be in #4B9E98 , currently `and earn 14.03% APY`

**Comment by pimato** [comment]
Replace current green with: #4B9E98 
Bildschirmfoto 2021-01-13 um 22 24 56

**Comment by pimato** [comment]
`earn 0.00% APY` should have a font-weight: 500

**Comment by pimato** [comment]
Circle button should be inactive on default!

**Comment by pimato** [comment]
Circle icon is not 40x40, currently:
Bildschirmfoto 2021-01-13 um 22 27 29

Should:
Bildschirmfoto 2021-01-13 um 22 29 05

**Comment by pimato** [comment]
Circle default styling is not correct, please see: 
https://www.figma.com/file/R056tSYMjvfwaFidEBItYy/Widgets-V2?node-id=1209%3A141

File: app/src/services/cpk.ts (hunk 8 above)
Comment (pRoy24) [reply app/src/services/cpk.ts:377]: Yes, This was the case during an interim commit when I incorrectly resolved a merge conflict after rebasing to master. Should be fixed now :)

File: app/src/components/market/common/add_compound_service/index.tsx (hunk 9 above)
Comment (pRoy24) [reply app/src/components/market/common/add_compound_service/index.tsx:126]: Great catch thanks I had not tested switching networks. Should be fixed now.

**Comment by pRoy24** [comment]
Resolved the css issues.
Fixed the issue with the app crashing in sokol and xDai.
Made some optimizations so that the interaction is smoother.
Rebased to master.

**Comment by pRoy24** [comment]
> on funding page loaded, app freezes

Hey @pimato Do you mean the app stops responding or are the interactions slower?
I haven't run into this issue.

The predicted amounts change in the display will be slightly slower than normal display updates since it needs to do the additional steps of getting the exchange rate and converting the amounts from baseToken -> cToken and vice versa.
I have made some optimizations so the interaction should be smoother now, let me know if during your testing you encounter the page freezing issue again.

**Comment by pimato** [comment]
just played around with it on Rinkeby and I am getting an error after clicking on circle button to enable it:
Bildschirmfoto 2021-01-15 um 16 35 19

Steps to reproduce:

1) Connect to Rinkeby 
2) Create Market 
3) Configure Market:
Bildschirmfoto 2021-01-15 um 16 37 04
4) Click Continue
5) Click enable Compound circle button
6) BUG: App crashes

**Comment by pRoy24** [comment]
> just played around with it on Rinkeby and I am getting an error after clicking on circle button to enable it:
> Bildschirmfoto 2021-01-15 um 16 35 19
> 
> Steps to reproduce:
> 
> 1. Connect to Rinkeby
> 2. Create Market
> 3. Configure Market:
> 
> Bildschirmfoto 2021-01-15 um 16 37 04
> 
> 4) Click Continue 5) Click enable Compound circle button 6) BUG: App crashes

You need to run Yarn install since it uses Big.js version 6.0.0 for the prec function

**Comment by pimato** [comment]
Hey @pRoy24 , once I enable compound the pool number is jumping around:
Bildschirmfoto 2021-01-18 um 13 13 11

**Comment by pimato** [comment]
Every market info component here has a svg in 24x24. The compound svg you included is 16x16:
Bildschirmfoto 2021-01-18 um 13 17 55

please use this svg:
[comp.svg.zip](https://github.com/protofire/omen-exchange/files/5830122/comp.svg.zip)

**Comment by pimato** [comment]
Please remove the styling if the circle button is active, currently if you hover over it you see the border is popping up:
Bildschirmfoto 2021-01-18 um 13 13 11

**Comment by pimato** [comment]
If you hover over the compound info the icon is turning blue, which is not according my design specification. Currently:
Bildschirmfoto 2021-01-18 um 13 21 37

Should:
Bildschirmfoto 2021-01-18 um 13 23 39

onHover color: #00897B

**Comment by pimato** [comment]
Just looking at the price of the outcomes here:
Bildschirmfoto 2021-01-18 um 13 27 24

I believe those prices are in cDAI and not in DAI right?

**Comment by pimato** [comment]
Looks like after I done a swap, the form is not resetting which is our default.

Doing a swap:
Bildschirmfoto 2021-01-18 um 13 31 05

after the swap is done, the inputfield is still filled with the number 2 and I see order details:
Bildschirmfoto 2021-01-18 um 13 31 54

It should reset the order form as we do on default.

**Comment by pimato** [comment]
> Please remove the styling if the circle button is active, currently if you hover over it you see the border is popping up:
> Bildschirmfoto 2021-01-18 um 13 13 11

it looks like now the the border got removed when holding on-click:
Bildschirmfoto 2021-01-19 um 09 27 24

**Comment by pimato** [comment]
The compound icon is still not correct, currently:
Bildschirmfoto 2021-01-19 um 10 14 33

should:
Bildschirmfoto 2021-01-19 um 10 13 40

**Comment by pimato** [comment]
> If you hover over the compound info the icon is turning blue, which is not according my design specification. Currently:
> Bildschirmfoto 2021-01-18 um 13 21 37
> 
> Should:
> Bildschirmfoto 2021-01-18 um 13 23 39
> 
> onHover color: #00897B

you changed the normal color to #00897B, but you actually should use #00897B as the hover color for the icon and the text.

**Review by pimato — CHANGES_REQUESTED** [review]
see comments.

File: app/src/components/market/common/add_compound_service/index.tsx [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  -webkit-box-align: center;
+`
+
+const ServiceTextWrapper = styled.div`
+  width: 90%;
+`
+
+const ServiceCheckWrapper = styled.div`
+  width: 10%;
+  color: transparent;
```
Comment (Mi-Lan) [inline app/src/components/market/common/add_compound_service/index.tsx:78]: I dont think defining static values to width here is neccesary. I would suggest just adding 
justify-content: space-between; to the parent element that is 100% and omitting width for these two values. The reason for this is because 10% is slightly more than 40px and creates this unnecessary margin...
Screenshot 2021-01-20 at 23 28 34

File: app/src/util/networks.ts [hunk 11]
```diff
@@ -260,13 +260,66 @@ export const getInfuraUrl = (networkId: number): string => {
 export const knownTokens: { [name in KnownToken]: KnownTokenData } = {
   cdai: {
     symbol: 'cDAI',
-    decimals: 18,
```
Comment (Mi-Lan) [inline app/src/util/networks.ts:263]: Screenshot 2021-01-20 at 23 31 04
Would be good to have consistent letter formatting for both currencies. To make them both capital or with first letter Uppercase (which is according to design...

**Review by Mi-Lan — CHANGES_REQUESTED** [review]
Small changes...

File: app/src/util/networks.ts (hunk 11 above)
Comment (pRoy24) [reply app/src/util/networks.ts:263]: Yup, makes sense refactored to include the change.

File: app/src/components/market/common/add_compound_service/index.tsx (hunk 10 above)
Comment (pRoy24) [reply app/src/components/market/common/add_compound_service/index.tsx:78]: Fixed
